# HG changeset patch
# User Mike Miller <mtmiller@octave.org>
# Date 1468620913 25200
#      Fri Jul 15 15:15:13 2016 -0700
# Node ID 64f82a8b46e130dc2f9822426dd826d9a732b620
# Parent  46ce9eb019a675769015ae21cba39ec5aeac849b
config.h: install a stub replacement with a deprecation warning (bug #41027)

* Makefile.am (nodist_octinclude_HEADERS): Remove autoconf-generated config.h
from the list of header files to install.
* deprecated-config.h: New file to replace config.h in the public API.  Use a
different name to not shadow config.h during the build.
* libinterp/module.mk (install-deprecated-config-h,
uninstall-deprecated-config-h): New targets to install and uninstall it.

diff --git a/Makefile.am b/Makefile.am
--- a/Makefile.am
+++ b/Makefile.am
@@ -305,17 +305,16 @@ DISTCLEANFILES += \
 MAINTAINERCLEANFILES += \
   $(BUILT_DISTFILES) \
   ChangeLog \
   HG-ID
 
 CONFIG_FILES = @ac_config_headers@ @ac_config_files@
 
 nodist_octinclude_HEADERS += \
-  config.h \
   oct-conf-post.h \
   octave-config.h
 
 ALL_LOCAL_TARGETS += \
   $(OCTAVE_INTERPRETER_TARGETS) \
   .gdbinit \
   $(DOC_TARGETS)
 
diff --git a/libinterp/deprecated-config.h b/libinterp/deprecated-config.h
new file mode 100644
--- /dev/null
+++ b/libinterp/deprecated-config.h
@@ -0,0 +1,30 @@
+/*
+
+Copyright (C) 2016 Mike Miller
+
+This file is part of Octave.
+
+Octave is free software; you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation; either version 3 of the License, or (at your
+option) any later version.
+
+Octave is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with Octave; see the file COPYING.  If not, see
+<http://www.gnu.org/licenses/>.
+
+*/
+
+#if ! defined (octave_config_h)
+#define octave_config_h 1
+
+#include "octave-config.h"
+
+#warning "config.h has been deprecated; there is no need to include it directly"
+
+#endif
diff --git a/libinterp/module.mk b/libinterp/module.mk
--- a/libinterp/module.mk
+++ b/libinterp/module.mk
@@ -283,19 +283,19 @@ libinterp/DOCSTRINGS: $(LIBINTERP_DEFUN_
 	$(call move_if_change_rule,libinterp/DOCSTRINGS-t,$@)
 
 OCTAVE_INTERPRETER_TARGETS += \
   $(OCT_FILES) \
   $(DLDFCN_PKG_ADD_FILE)
 
 DIRSTAMP_FILES += libinterp/$(octave_dirstamp)
 
-install-data-hook: install-oct install-built-in-docstrings
+install-data-hook: install-oct install-built-in-docstrings install-deprecated-config-h
 
-uninstall-local: uninstall-oct uninstall-built-in-docstrings
+uninstall-local: uninstall-oct uninstall-built-in-docstrings uninstall-deprecated-config-h
 
 if AMCOND_ENABLE_DYNAMIC_LINKING
 install-oct:
 	$(MKDIR_P) $(DESTDIR)$(octfiledir)
 	if [ -n "`cat $(DLDFCN_PKG_ADD_FILE)`" ]; then \
 	  $(INSTALL_DATA) $(DLDFCN_PKG_ADD_FILE) $(DESTDIR)$(octfiledir)/PKG_ADD; \
 	fi
 	cd $(DESTDIR)$(octlibdir) && \
@@ -328,16 +328,25 @@ install-built-in-docstrings:
 	  if test -f $$f; then d=; else d="$(srcdir)/"; fi; \
 	  $(INSTALL_DATA) "$$d$$f" $(DESTDIR)$(octetcdir)/built-in-docstrings
 .PHONY: install-built-in-docstrings
 
 uninstall-built-in-docstrings:
 	rm -f $(DESTDIR)$(octetcdir)/built-in-docstrings
 .PHONY: uninstall-built-in-docstrings
 
+install-deprecated-config-h: libinterp/deprecated-config.h
+	$(MKDIR_P) $(DESTDIR)$(octincludedir)
+	$(INSTALL_DATA) $< $(DESTDIR)$(octincludedir)/config.h
+.PHONY: install-deprecated-config-h
+
+uninstall-deprecated-config-h:
+	rm -f $(DESTDIR)$(octincludedir)/config.h
+.PHONY: uninstall-deprecated-config-h
+
 EXTRA_DIST += $(libinterp_EXTRA_DIST)
 
 libinterp_CLEANFILES += \
   $(DLDFCN_PKG_ADD_FILE) \
   $(LIBINTERP_BUILT_NODISTFILES) \
   $(LIBINTERP_TST_FILES) \
   $(OCT_FILES) \
   libinterp/corefcn/oct-tex-parser.output \
